compiler/rustc_trait_selection/src/traits/const_evaluatable.rs

```diff
@@ -200,22 +199,37 @@ fn satisfied_from_param_env<'tcx>(
         }
     }
 
+    let mut single_match: Option<Result<ty::Const<'tcx>, ()>> = None;
+
     for pred in param_env.caller_bounds() {
         match pred.kind().skip_binder() {
             ty::PredicateKind::ConstEvaluatable(ce) => {
                 let b_ct = tcx.expand_abstract_consts(ce);
-                let mut v = Visitor { ct, infcx, param_env };
-                let result = b_ct.visit_with(&mut v);
-
-                if let ControlFlow::Break(()) = result {
-                    debug!("is_const_evaluatable: yes");
-                    return true;
+                let mut v = Visitor { ct, infcx, param_env, single_match: None };
+                let _ = b_ct.visit_with(&mut v);
+                if let Some(inner) = v.single_match {
+                    single_match = if single_match.is_none() { Some(inner) } else { Some(Err(())) };
                 }
             }
             _ => {} // don't care
         }
     }
 
+    if let Some(c) = single_match {
+        if let Ok(c) = c {
+            let is_ok = infcx
+                .commit_if_ok(|_| {
+                    let ocx = ObligationCtxt::new_in_snapshot(infcx);
+                    assert!(ocx.eq(&ObligationCause::dummy(), param_env, c.ty(), ct.ty()).is_ok());
+                    assert!(ocx.eq(&ObligationCause::dummy(), param_env, c, ct).is_ok());
+                    if ocx.select_all_or_error().is_empty() { Ok(()) } else { Err(()) }
+                })
+                .is_ok();
+            assert!(is_ok);
+        }
+        return true;
```

Comment:
You shouldn't return true if you havent rerun the `ocx.eq` and `ocx.select_all_or_error` this will lead to us considering stuff to be evaluatable that actually isnt:
```rust
#![feature(generic_const_exprs)]
#![allow(incomplete_features)]

fn foo<const N: usize, const M: usize>() -> [(); N + 2]
where
    [(); N + 1]:,
    [(); M + 1]:,
{
    bar()
}

fn bar<const N: usize>() -> [(); N]
where
    [(); N + 1]:,
{
    [(); N]
}

fn main() {}
```
this ICEs with your PR and will be fixed by doing `if let Some(Ok(c)) = single_match {` instead, so that you only return true if there was a single match and you have constrained all the infer vars

Reply:
While this change does fix this specific case, it does make other cases not evaluatable when they should be.
I'm not sure there's an easy way to fix it without changing more in other places? I tried to return `MentionsInfer` but that didn't seem to work